module5/lessons/Week3/Authentication.md

```diff
@@ -0,0 +1,289 @@
+* What other relationships can you label?
+
+</section>
+
+### Experiment!
+
+Now that we have some aspects of our application set up, play around with the following:
+* Register a new user
+* Add some messages
+* Inspect the records created in pgAdmin
+
+We are getting closer to our handrolled functionality, but we are still missing a few things.
+
+<section class='call-to-action' markdown='1'>
+
+With a partner, discuss the functionality that is missing.  Be ready to share out!
+
+</section>
+
+### Get Current User
+
+For many functions that we want in our MessageBoard application, we will need access to the current user (the user who is currently logged in).
+
+<section class='call-to-action' markdown='1'>
+
+In the application, when a user is logged in, we can see their name in the upper right corner of the page.  Explore the code to figure out how we can get information about a current user into a specific view.
+
+Hint: Think about how we got the user's email address into the nav-bar!
+
+Spend at least 5 minutes playing around before checking out the answer below.
+
+</section>
+
+<section class='answer' markdown='1'>
+
+### the User
+
+If we take a look at our `/Views/Shared/_LoginPartial`, we can see that we are using a combination of `SignInManager` and `UserManager` objects from the Identity framework to manage our current user.
+
+`SignInManager` is responsible for checking that there **is** someone signed in.
+
+`UserManager` is responsible for providing data about a `User`.  With the Identity framework, we have access to a `User` throughout the application.  We can use that object, and the UserManager to get the information we need about a current user.
+
+```html
+@using Microsoft.AspNetCore.Identity
+@using IdentityMessageBoard.Models
+
+@inject SignInManager<ApplicationUser> SignInManager
+@inject UserManager<ApplicationUser> UserManager
+
+<ul class="navbar-nav">
+@if (SignInManager.IsSignedIn(User))
+{
+    <li class="nav-item">
+        <a id="manage" class="nav-link text-dark" asp-area="Identity" asp-page="/Account/Manage/Index" title="Manage">Hello @UserManager.GetUserName(User)!</a>
+    </li>
+    <li class="nav-item">
+        <form id="logoutForm" class="form-inline" asp-area="Identity" asp-page="/Account/Logout" asp-route-returnUrl="@Url.Action("Index", "Home", new { area = "" })">
+            <button id="logout" type="submit" class="nav-link btn btn-link text-dark">Logout</button>
+        </form>
+    </li>
+}
+else
+{
+    <li class="nav-item">
+        <a class="nav-link text-dark" id="register" asp-area="Identity" asp-page="/Account/Register">Register</a>
+    </li>
+    <li class="nav-item">
+        <a class="nav-link text-dark" id="login" asp-area="Identity" asp-page="/Account/Login">Login</a>
+    </li>
+}
+</ul>
+```
+
+</section>
+
+<section class='call-to-action' markdown='1'>
+
+**With a Partner**  
+Use what we have learned about accessing current user data to implement the following user story:
+
```

Comment:
This seems like a pretty tricky task, I wonder if it would flow more smoothly to break this into two parts. Pseudocode/plan out the changes you will need to make. Then come back together to talk about it and then break again to implement.

Reply:
Totally - good call!